haskell.packages.ghc96: bump memory and parallel

TristanCacqueray opened the pull request:
###### Description of changes

This change enables using the memory and parallel package.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

TristanCacqueray committed:
haskell.packages.ghc96.memory: bump version

This change enables using the memory package.

TristanCacqueray committed:
haskell.packages.ghc96.parallel: bump version

TristanCacqueray commented:
Please let me know if I forgot something to make your life easier :)

cdepillabout commented:
@ofborg build haskell.packages.ghc96.memory haskell.packages.ghc96.parallel

cdepillabout commented:
@TristanCacqueray Thanks for sending this!

When asking ofborg to build this, it seems to be failing on `haskell.packages.ghc96.foundation`.  Are you not seeing this when you build locally?

TristanCacqueray commented:
You are welcome, thank you for testing the build!

So we need https://github.com/haskell-foundation/foundation/pull/572 . I tried to use a fetchPatch instead of cloning the branch, but that doesn't seems to work because of the subdir prefix:

```nix
{
        foundation = pkgs.haskell.lib.overrideCabal (hpPrev.foundation) {
          patches = [
            (pkgs.fetchpatch { # fetchpatch2 doesn't help
              url =
                "https://patch-diff.githubusercontent.com/raw/haskell-foundation/foundation/pull/572.patch";
              sha256 = "sha256-xn73YzxQNRcxYXw0R+QTQUgeSnWlBcO/BdPRS2bqmfI=";
              stripLen = 2;
              # includes = [ "foundation/*" ];
            })
          ];
        };
}
```

Otherwise, could we do the following?
```nix
{
        foundation = pkgs.haskell.lib.overrideCabal (hpPrev.foundation) {
          # https://github.com/haskell-foundation/foundation/pull/572
          src = let
            foundation-src = pkgs.fetchFromGitHub {
              owner = "Topsii";
              repo = "foundation";
              rev = "d3136f4bb8b69e273535352620e53f2196941b35";
              sha256 = "sha256-Kxf+yF8VrKVezQVBzrzrN6tOaoxl9f4Bub2IvalvMdA=";
	    };
	  in "${foundation-src}/foundation";
	};
}
```

cdepillabout commented:
`fetchPatch` is generally preferable to overriding the source of a package.  This is because when a package is updated on Hackage, `fetchPatch` will generally start failing, which indicates to us we can remove the patch.  But if you override the source, then we don't get any indication that the source override should be removed.

There are also a bunch of options you can pass to `fetchPatch` to pull out exactly the files you want.  Here's an example.  Would something like this help?

https://github.com/NixOS/nixpkgs/blob/2d30d94721f14b4a799743a8955b4e2ed4b50576/pkgs/tools/text/difftastic/default.nix#L26

If this doesn't work, then you might be able to override the source of the package, but use a version assertion like the following:

https://github.com/NixOS/nixpkgs/blob/49d835a29f6b4f033aaa2a6fb303bda422e28650/pkgs/development/haskell-modules/configuration-common.nix#L2116

sternenseemann commented:
@TristanCacqueray I think `stripLen = 1;` alone should do the trick. `patch` doesn't care if the first component is `a` and `b` or `foundation` and `foundation`.

TristanCacqueray committed:
haskell.packages.ghc96.foundation: add build patch

TristanCacqueray commented:
It seemed like without changing the name, the options like `stripLen` or `extraPrefix` are ignored. I got this last commit to build using such test:

```nix
(import ./. {}).haskell.packages.ghc961.ghcWithPackages (p: [ p.memory	p.parallel ]) 
```

cdepillabout commented:
@ofborg build haskell.packages.ghc96.memory haskell.packages.ghc96.parallel haskell.packages.ghc96.foundation

TristanCacqueray commented:
I wonder if we should avoid pull request patch because the content might change if the PR is updated. Should I pick the commit from the source branch instead?

sternenseemann commented on pkgs/development/haskell-modules/configuration-ghc-9.6.x.nix:
```diff
@@ -142,11 +144,27 @@ self: super: {
+      (pkgs.fetchpatch {
+        name = "foundation-pr-572.patch";
+        url =
+          "https://patch-diff.githubusercontent.com/raw/haskell-foundation/foundation/pull/572.patch";
```
Usually better to use the content-addressed commit instead of the mutable PR reference, so the `sha256` is never invalidated.


```suggestion
          "https://github.com/haskell-foundation/foundation/commit/d3136f4bb8b69e273535352620e53f2196941b35.patch";
```

sternenseemann commented on pkgs/development/haskell-modules/configuration-ghc-9.6.x.nix:
```diff
@@ -142,11 +144,27 @@ self: super: {
+          "https://patch-diff.githubusercontent.com/raw/haskell-foundation/foundation/pull/572.patch";
+        sha256 = "sha256-oPadhQdCPJHICdCPxn+GsSQUARIYODG8Ed6g2sK+eC4=";
+        stripLen = 1;
+      })
+    ];
+  };
```
To match above.

```suggestion
      })
    ] (super.foundation);
```

maralorn commented:
> I wonder if we should avoid pull request patch because the content might change if the PR is updated. Should I pick the commit from the source branch instead?

Yeah, always try to sanitize the commit path by removing the pull request from the path.

TristanCacqueray commented:
@maralorn yeah, I was wondering why changing any attributes always resulted in the same patch, until I saw difftastic example and added a name :)

@sternenseemann that looks much better, thanks! I squashed the update onto the existing commit, I hope that's ok.

sternenseemann commented:
@ofborg build haskell.packages.ghc961.memory haskell.packages.ghc961.parallel haskell.packages.ghc961.foundation

TristanCacqueray commented:
@sternenseemann That's good to know. For what it's worth, I had cabal working two weeks ago with [this flake](https://github.com/podenv/hspkgs/blob/4662e1524d9ad208059b7c0458d57fad98e5ba41/flake.nix#L383-L494). I didn't knew about `configuration-ghc-9.6.x.nix`, next time I'll make the override here instead, it's much easier :)

Thank you for the review!

sternenseemann commented:
`cabal-install` was the goal I worked towards for initial support and already works on haskell-updates.
